### PR TITLE
4.0.0 release

### PR DESCRIPTION
**Pull request by jjasghar**
This has a ton of changes and feature requests.

I'm betting it's a Major release, but i'd like some feedback.

### CODE REVIEW

File: libraries/matchers.rb [hunk 1]
```diff
@@ -54,4 +54,16 @@ def enable_rabbitmq_plugin(resource_name)
   def disable_rabbitmq_plugin(resource_name)
     ChefSpec::Matchers::ResourceMatcher.new(:rabbitmq_plugin, :disable, resource_name)
   end
+
+  def join_cluster(resource_name)
+    ChefSpec::Matchers::ResourceMatcher.new(:rabbitmq_cluster, :join, resource_name)
+  end
+
+  def set_cluster_name(resource_name) # rubocop:disable AccessorMethodName
+    ChefSpec::Matchers::ResourceMatcher.new(:rabbitmq_cluster, :change_cluster_node_type, resource_name)
```
Comment (kramvan1) [inline libraries/matchers.rb]: Humm, this looks like a typo, should be  :set_cluster_name  as 2nd parm?

File: libraries/matchers.rb [hunk 2]
```diff
@@ -54,4 +54,16 @@ def enable_rabbitmq_plugin(resource_name)
   def disable_rabbitmq_plugin(resource_name)
     ChefSpec::Matchers::ResourceMatcher.new(:rabbitmq_plugin, :disable, resource_name)
   end
+
+  def join_cluster(resource_name)
+    ChefSpec::Matchers::ResourceMatcher.new(:rabbitmq_cluster, :join, resource_name)
```
Comment (kramvan1) [inline libraries/matchers.rb]: maybe :join_cluster  as 2nd parm to be consistent here?

File: providers/cluster.rb [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+# limitations under the License.
+#
+
+include Chef::Mixin::ShellOut
+
+use_inline_resources
+
+# Get ShellOut
+def get_shellout(cmd)
+  sh_cmd = Mixlib::ShellOut.new(cmd)
```
Comment (kramvan1) [inline providers/cluster.rb]: Is there any concern here with the default 5min timeout for ShellOut?

File: recipes/default.rb [hunk 4]
```diff
@@ -32,6 +32,16 @@ class Chef::Resource # rubocop:disable all
   # logrotate is a package dependency of rabbitmq-server
   package 'logrotate'
 
+  # dpkg, imma let you finish but don't start services automatically
+  # https://jpetazzo.github.io/2013/10/06/policy-rc-d-do-not-start-services-automatically/
```
Comment (kramvan1) [inline recipes/default.rb]: Is this hack same on all platforms?

File: spec/cluster_spec.rb [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+
+require 'spec_helper'
+
+describe 'rabbitmq::cluster' do
+  let(:runner) { ChefSpec::ServerRunner.new(REDHAT_OPTS) }
+  let(:node) { runner.node }
+  let(:chef_run) do
+    runner.converge(described_recipe)
+  end
+
```
Comment (kramvan1) [inline spec/cluster_spec.rb]: pretty think specs for all that lwrp code, hopefully more will be comming for this new support.

File: recipes/default.rb (hunk 4 above)
Comment (jjasghar) [inline recipes/default.rb]: It's in the debian family section. Line 31

**Comment by kramvan1** [comment]
Added some minor comments.  With this new lwrp, certainly seems like this could be a 4.x release, but if this new lwrp is not used in default path, it could be justified as just a minor enhancement.

**Comment by jjasghar** [comment]
With rolling in the clustering cookbook i feel justified on the 4.x release, if i can get someone other than me verify that `kitchen` and the `specs` pass i'm ready to merge this.

File: providers/cluster.rb (hunk 3 above)
Comment (cheeseplus) [inline providers/cluster.rb]: There is always a concern for timeouts on shellout, especially for cluster operations. We should probably note that or make it configurable.

File: libraries/matchers.rb (hunk 1 above)
Comment (jjasghar) [inline libraries/matchers.rb]: Nope, it looks like it's referenced as `:change_cluster_node_type`

File: libraries/matchers.rb (hunk 2 above)
Comment (jjasghar) [inline libraries/matchers.rb]: :-1:

File: recipes/default.rb (hunk 4 above)
Comment (cheeseplus) [inline recipes/default.rb]: Correct, the hack will only apply to debian platform_families although other commits in that PR do fix the root issue on both RHEL and Debian platform_families

File: providers/cluster.rb (hunk 3 above)
Comment (jjasghar) [inline providers/cluster.rb]: that should be for `4.1.0` :)

**Pull request merged**